libkbfs/block_ops_constrained.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+// BlockOpsConstrained implements the BlockOps interface by relaying
+// requests to a delegate BlockOps, but it delays all Puts by
+// simulating a bottleneck of the given bandwidth.
+type BlockOpsConstrained struct {
+	BlockOps
+	bwKBps int
+	lock   sync.Mutex
+}
+
+// NewBlockOpsConstrained constructs a new BlockOpsConstrained.
+func NewBlockOpsConstrained(delegate BlockOps, bwKBps int) *BlockOpsConstrained {
+	return &BlockOpsConstrained{
+		BlockOps: delegate,
+		bwKBps:   bwKBps,
+	}
+}
+
+// Put implements the BlockOps interface for BlockOpsConstrained.
+func (b *BlockOpsConstrained) Put(ctx context.Context, md *RootMetadata,
+	blockPtr BlockPointer, readyBlockData ReadyBlockData) error {
+	if b.bwKBps > 0 {
+		b.lock.Lock()
```

Comment:
Nit: Mind decomping out the locked part in a helper function so you can do `defer Unlock()`?